File: pyg4ometry/pygeometry/cpgeant4/solid/Tet.cpp
#include "Tet.h"
#include <vector>

CSG* CSGMesh::ConstructTet(Vector* anchor,Vector* p2,Vector* p3,Vector* p4,bool degeneracyFlag){
  Vertex* vert_ancr = new Vertex(anchor,NULL);
  std::vector<Vertex*> vert_base(3);
  vert_base[0] = new Vertex(p2,NULL);
  vert_base[1] = new Vertex(p3,NULL);
  vert_base[2] = new Vertex(p4,NULL);

  std::vector<Polygon*> polygons(4);

  std::vector<Vertex*> vtx1(3);
  vtx1[0] = vert_base[2]; vtx1[1] = vert_base[1]; vtx1[2] = vert_base[0];
  polygons[0] = new Polygon(vtx1,NULL);

  std::vector<Vertex*> vtx2(3);
  vtx2[0] = vert_base[1]; vtx2[1] = vert_ancr; vtx2[2] = vert_base[0];
  polygons[1] = new Polygon(vtx2,NULL);

  std::vector<Vertex*> vtx3(3);
  vtx3[0] = vert_base[2]; vtx3[1] = vert_ancr; vtx3[2] = vert_base[1];
  polygons[2] = new Polygon(vtx3,NULL);

  std::vector<Vertex*> vtx4(3);
  vtx4[0] = vert_base[0]; vtx4[1] = vert_ancr; vtx4[2] = vert_base[2];
  polygons[3] = new Polygon(vtx4,NULL);

  return CSG::fromPolygons(polygons);
}

File: src/pyg4ometry/geant4/solid/Tet.py
from ... import config as _config

from .SolidBase import SolidBase as _SolidBase

if _config.meshing == _config.meshingType.pycsg:
    from ...pycsg.core import CSG as _CSG
    from ...pycsg.geom import Vector as _Vector
    from ...pycsg.geom import Vertex as _Vertex
    from ...pycsg.geom import Polygon as _Polygon
elif _config.meshing == _config.meshingType.cgal_sm:
    from ...pycgal.core import CSG as _CSG
    from ...pycgal.geom import Vector as _Vector
    from ...pycgal.geom import Vertex as _Vertex
    from ...pycgal.geom import Polygon as _Polygon

import numpy as _np
import logging as _log
from copy import deepcopy as _dc


class Tet(_SolidBase):
    """
    Constructs a tetrahedra.

    :param name:           of the solid
    :param anchor:         point 1 (anchor point)
    :type anchor:          list
    :param p2:             point 2
    :type p2:              list
    :param p3:             point 3
    :type p3:              list
    :param p4:             point 4
    :type p4:              list
    :param registry:       for storing solid
    :type registry:        Registry
    :param lunit:          length unit (nm,um,mm,m,km) for solid
    :type lunit:           str

    :param degeneracyFlag: bool, indicates degeneracy of points
    """

    def __init__(
        self,
        name,
        anchor,
        p2,
        p3,
        p4,
        registry,
        lunit="mm",
        degeneracyFlag=False,
        addRegistry=True,
    ):
        super().__init__(name, "Tet", registry)

        self.anchor = anchor
        self.p2 = p2
        self.p3 = p3
        self.p4 = p4
        self.lunit = lunit
        self.degen = degeneracyFlag

        self.dependents = []

        self.varNames = ["anchor", "p2", "p3", "p4"]
        self.varUnits = ["lunit", "lunit", "lunit", "lunit"]

        if addRegistry:
            registry.addSolid(self)  # Always need the registry

    def __repr__(self):
        return f"Tet : {self.name} Vertexes: {self.anchor!s}, {self.p2!s}, {self.p3!s}, {self.p4!s}"

    def __str__(self):
        return f"Tet : name={self.name} Vertexes: a={self.anchor!s}, p2={self.p2!s}, p3={self.p3!s}, p4={self.p4!s}"

    def mesh(self):
        _log.info("tet.pycsgmesh> antlr")

        from ...gdml import Units as _Units

        luval = _Units.unit(self.lunit)

        anchor = [val * luval for val in self.evaluateParameter(self.anchor)]
        p2 = [val * luval for val in self.evaluateParameter(self.p2)]
        p3 = [val * luval for val in self.evaluateParameter(self.p3)]
        p4 = [val * luval for val in self.evaluateParameter(self.p4)]

        _log.info("tet.pycsgmesh> mesh")
        vert_ancr = _Vertex(_Vector(p4[0], p4[1], p4[2]))
        base = [anchor, p2, p3]
        vert_base = []

        for i in range(len(base)):
            vert_base.append(_Vertex(_Vector(base[i][0], base[i][1], base[i][2])))

        mesh = _CSG.fromPolygons(
            [
                _Polygon([vert_base[2], vert_base[1], vert_base[0]]),
                _Polygon([vert_base[1], vert_ancr, vert_base[0]]),
                _Polygon([vert_base[2], vert_ancr, vert_base[1]]),
                _Polygon([vert_base[0], vert_ancr, vert_base[2]]),
            ]
        )

        return mesh
